x509/golang.go

```diff
@@ -112,8 +118,115 @@ func marshalSubject(cn string, ou string) pkix.Name {
 	return pkix.Name{ExtraNames: pkixAttrTypeValue}
 }
 
-func CreateFactoryCa(ou string) string {
-	priv := genAndSaveKey(FactoryCaKeyFile)
+func genAndSaveKeyToHsmNoRet(hsm HsmInfo, id []byte, label []byte) {
+	p := pkcs11.New(hsm.Module)
+	err := p.Initialize()
+	subcommands.DieNotNil(err)
+
+	defer p.Destroy()
+	defer p.Finalize()
+
+	slots, err := p.GetSlotList(true)
+	subcommands.DieNotNil(err)
+
+	var matchedSlotId uint = math.MaxUint
+	for _, slotId := range slots {
+		token, err := p.GetTokenInfo(slotId)
+		if err != nil {
+			continue
+		}
+		if token.Label == hsm.TokenLabel {
+			matchedSlotId = slotId
+			break
+		}
+	}
+
+	if matchedSlotId == math.MaxUint {
+		subcommands.DieNotNil(errors.New("PKCS#11: Unable to find Token" + hsm.TokenLabel))
+	} else {
+		fmt.Printf("PKCS#11: Found Token %s at SlotID %d\n", hsm.TokenLabel, matchedSlotId)
+	}
+
+	session, err := p.OpenSession(matchedSlotId, pkcs11.CKF_SERIAL_SESSION|pkcs11.CKF_RW_SESSION)
+	subcommands.DieNotNil(err)
+	defer p.CloseSession(session)
+
+	err = p.Login(session, pkcs11.CKU_USER, hsm.Pin)
+	subcommands.DieNotNil(err)
+	defer p.Logout(session)
+
+	oidNamedCurveP256 := asn1.ObjectIdentifier{1, 2, 840, 10045, 3, 1, 7}
+	marshaledOID, err := asn1.Marshal(oidNamedCurveP256)
+
+	publicKeyTemplate := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, pkcs11.CKK_EC),
+		pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_PUBLIC_KEY),
+		pkcs11.NewAttribute(pkcs11.CKA_TOKEN, true),
+		pkcs11.NewAttribute(pkcs11.CKA_VERIFY, true),
+		pkcs11.NewAttribute(pkcs11.CKA_EC_PARAMS, marshaledOID),
+		pkcs11.NewAttribute(pkcs11.CKA_DERIVE, true),
+
+		pkcs11.NewAttribute(pkcs11.CKA_ID, id),
+		pkcs11.NewAttribute(pkcs11.CKA_LABEL, label),
+	}
+	privateKeyTemplate := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, pkcs11.CKK_EC),
+		pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_PRIVATE_KEY),
+		pkcs11.NewAttribute(pkcs11.CKA_TOKEN, true),
+		pkcs11.NewAttribute(pkcs11.CKA_PRIVATE, true),
+		pkcs11.NewAttribute(pkcs11.CKA_SIGN, true),
+
+		pkcs11.NewAttribute(pkcs11.CKA_ID, id),
+		pkcs11.NewAttribute(pkcs11.CKA_LABEL, label),
+
+		pkcs11.NewAttribute(pkcs11.CKA_EXTRACTABLE, false),
+		pkcs11.NewAttribute(pkcs11.CKA_SENSITIVE, true),
+
+		pkcs11.NewAttribute(pkcs11.CKA_DERIVE, true),
+	}
+	_, _, err = p.GenerateKeyPair(session,
+		[]*pkcs11.Mechanism{pkcs11.NewMechanism(pkcs11.CKM_EC_KEY_PAIR_GEN, nil)},
+		publicKeyTemplate, privateKeyTemplate)
+	subcommands.DieNotNil(err)
+}
+
+func getKeyHsm(hsm HsmInfo, id []byte) crypto11.Signer {
+	cfg := crypto11.Config{
+		Path:        hsm.Module,
+		TokenLabel:  hsm.TokenLabel,
+		Pin:         hsm.Pin,
+		MaxSessions: 0,
+	}
+
+	ctx, err := crypto11.Configure(&cfg)
+	subcommands.DieNotNil(err)
+
+	privKey, err := ctx.FindKeyPair(id, nil)
+	subcommands.DieNotNil(err)
```

Comment:
Just asking... So, pkcs11 interface is too low-level to conveniently work with it, right?
That's why we need ThalesIgnite: to call two functions instead of writing 100500 lines of obscure code ourselves, right?

Reply:
Makes sense.